### PR TITLE
Addressing R CMD check notes due to tidyeval syntax

### PR DESCRIPTION
**Pull request by gvegayon**
- ~Adds `+ file LICENSE` to `DESCRIPTION`~ Adds `LICENSE` to the `.Rbuildignore` file (R convention).
- Moves `withr` to `Suggests`, as it is only used during testing.
- Adds the `rlang::.data` imports to address masking notes.
- Makes use of `.data$` for `dplyr` verbs when needed (see [here](https://dplyr.tidyverse.org/articles/in-packages.html)).
- Replaces expressions with strings in `dplyr` verbs when needed.
- ~Flags various variables to avoid warnings during R CMD check via `utils::globalVariables`. All variables in the list are those used in `tidybayes::spread_draws` calls (which uses expressions).~
- Adds some files/folders to `.Rbuildignore`.
- Modifies the `NEWS.md` document to follow R's convention (see also #118). 
- Ensures all `ggplot2` functions are imported and thus removes the use of `ggplot2::` (**NOTE: `dplyr::` is still heavily used).**

### CODE REVIEW

**Review by kaitejohnson — COMMENTED** [review]
One how did you do this all in a day. You rock. 

Two I will admit I hate this so much... I know the other option is `.data[variable]` and that's not much better...

@seabbs suggested that there is a way to use `autoglobals()` and we could also import the common functions to avoid having to use things like `ggplot2::ggplot()` and such everywhere (though I know there is currently lots of redundancy in this). I just would prefer the code be a bit more readable...

**Comment by gvegayon** [comment]
> One how did you do this all in a day. You rock.

Thank you!

> Two I will admit I hate this so much... I know the other option is `.data[variable]` and that's not much better...
> 
> @seabbs suggested that there is a way to use `autoglobals()` and we could also import the common functions to avoid having to use things like `ggplot2::ggplot()` and such everywhere (though I know there is currently lots of redundancy in this). I just would prefer the code be a bit more readable...

Yes, I don't like it either, but it's the way `dplyr` says it should be done. Just checked the [autoglobals alternative using roxygen tags](https://github.com/anthonynorth/roxyglobals); looks handy! One question: How reliable is that R package? We wouldn't be adding a dependency for the users, but we would be doing so for the developers. Perhaps avoid the package and use `utils::globalVariables()`? It would be pretty much the same thing!

**Comment by dylanhmorris** [comment]
> * Flags various variables to avoid warnings during R CMD check via `utils::globalVariables`. All variables in the list are those used in `tidybayes::spread_draws` calls (which uses expressions).

Thanks for this, @gvegayon 

For `tidybayes::spread_draws`, we can do calls of the form

```R
fit |> tidybayes::spread_draws(!!str2lang("variable name"))
```

See https://github.com/mjskay/tidybayes/issues/38#issuecomment-895281257

File: R/generate_simulated_data.R
Comment (dylanhmorris) [reply R/generate_simulated_data.R:421]: For consistency with how you're doing `rename` calls elsewhere

File: R/get_stan_data.R
Comment (dylanhmorris) [inline R/get_stan_data.R:243]:
```suggestion
```

File: R/get_stan_data.R
Comment (dylanhmorris) [inline R/get_stan_data.R:244]:
```suggestion
```

File: R/model_component_fwd_sim.R
Comment (dylanhmorris) [inline R/model_component_fwd_sim.R:372]:
```suggestion
      "lod_sewage" =
```

File: R/preprocessing.R
Comment (dylanhmorris) [inline R/preprocessing.R:57]:
```suggestion
         "lab_site_index" = dplyr::row_number()
```

File: R/preprocessing.R
Comment (dylanhmorris) [inline R/preprocessing.R:64]:
```suggestion
        dplyr::mutate("site_index" = dplyr::row_number()),
```

File: R/preprocessing.R
Comment (dylanhmorris) [inline R/preprocessing.R:182]:
```suggestion
      .data$n_data_points > !!threshold_n_dps
```

File: R/preprocessing.R
Comment (dylanhmorris) [inline R/preprocessing.R:210]:
```suggestion
          mean_conc = mean(.data[[conc_col_name]], na.rm = TRUE),
          std_conc = sd(.data[[conc_col_name]], na.rm = TRUE)
```

File: R/preprocessing.R
Comment (dylanhmorris) [inline R/preprocessing.R:216]:
```suggestion
      z_score_conc = (.data[[conc_col_name]] - .data$mean_conc) /
```

File: R/preprocessing.R
Comment (dylanhmorris) [inline R/preprocessing.R:230]:
```suggestion
            (abs(.data$z_score_rho_t_plus_1) >= !!rho_threshold) &
```

File: R/preprocessing.R
Comment (dylanhmorris) [inline R/preprocessing.R:229]:
```suggestion
          abs(.data$z_score_rho) >= !!rho_threshold &
```

File: R/preprocessing.R
Comment (dylanhmorris) [inline R/preprocessing.R:223]:
```suggestion
        abs(.data$z_score_conc) >= !!log_conc_threshold ~ 1,
```

**Review by dylanhmorris — COMMENTED** [review]
Thanks so much for this, @gvegayon. A number of things, all very small.

**Review by kaitejohnson — COMMENTED** [review]
Can we maybe have a quick chat before this gets merged in? I think I need to spend some time looking into the `autoglobals()` package/ what @gvegayon suggested with `utils::globalvariables` and understand the tradeoffs of that approach (which seem to just be a developer dependency?) over the `dplyr()` recommendation... 

I have a pretty strong preference to avoid resyntaxing the whole package this way. I know it doesn't materially change anything, but I personally find it super hard to follow and I think there is a chance that this deters users trying to understand what the code is doing. And since this is an R package for a very specific use case of a model rather than as say a framework for model development, I'm inclined to take some short cuts on optimal development, at least in this first release, to make it more approachable to people familiar with R/tidyverse syntax.

**Comment by seabbs** [comment]
This is a lot of work so well done @gvegayon but as discussed here I find this code nearly unreadable and the benefits seem very small. 

I think if we used the package namespace and imported our core tooling (i.e from dplyr, and ggplot2) this would help a lot with the `package::function` bloat. I see few if any downsides to doing that. 

The downside to not using `.data$var` is that as @gvegayon points out it is recommended practice and not doing so could add some small danger. However I think the risk from this is fairly marginal in most of this code base so would prefer we using globals (but I feel less strongly about this than the package::function who's removal would make the code much easier to read and so make this issue less). 

In terms of just manually assigning the globals. Yes, we could but it is exceptionally tedious. As it's a light dev dependency where any change will run through CI I don't think having it as a dev dependency is a serious issue. @athowes how have you found it?

**Comment by athowes** [comment]
I don't know that I understand the full context here but I've found using the [`roxyglobals`](https://github.com/anthonynorth/roxyglobals) easy enough.

About the other parts, it's not really my place to say / it's up to what you collectively decide, but if I were developing this package I would probably have reservations about the `pkg::` to everything and `.data$var` points. Perhaps that's just selfishness (it makes development less nice) but I think there are reasonable arguments to be made about code readability reducing the amount of error. With something like putting `ggplot2::` I might also question the extent to which this is ever likely to create bugs that aren't obvious / very easy to fix (I think that with some other more internal functions you might expect that an undetected namespace clash has more of a risk of creating a hard to see bug).

**Comment by kaitejohnson** [comment]
@gvegayon I'm inclined to give the `roxyglobals` a try (happy to do this in a separate PR and then compare the pros and cons). I think I saw that you had already started adding `@autoglobal` to some of the new functions you wrote?

**Comment by kaitejohnson** [comment]
Ok after discussing face-to-face, it seems like this is going to have to be the way to go in terms of the `.data[[]` and string calls, so will defer to @dylanhmorris on the requested changes. 

One remaining question I have is whether we could still do some amount of importing common package functions like `ggplot`, `aes`, `geom_line`, etc. or does that also always cause errors? It doesn't matter too much but I think it might help slightly with readability and is definitely hard to get into the habit of doing for things like ggplot

**Comment by dylanhmorris** [comment]
Yes, important to separate the question of importing any/many functions or other names from other packages from the question of how to handle column names in tidyeval functions. We can continue to import things like `ggplot2::ggplot` etc (and note that per my review this PR sticks with importing `ggplot2::aes`)

**Comment by dylanhmorris** [comment]
> .data[[] and string calls
Clarification that there are some places we'd be able to get away with tidy syntax but other places (dynamic column names) where we'll definitely need the `.data` pronoun: 

https://github.com/CDCgov/ww-inference-model/blob/7bcbebc4965b52759f9ba7b80c71f67f5b6cbc7b/R/preprocessing.R#L299

But given that:
- we'll be using `.data` of necessity in some places
- given that it's the tidyverse recommendation for packaged code as @gvegayon notes

I favor just using it throughout

**Comment by dylanhmorris** [comment]
In short: I support removing the namespace qualifications from this PR (e.g. revert `ggplot2::ggplot()` to `ggplot()`, but keeping the universal use of the `.data` pronoun. @kaitejohnson, @gvegayon et al does that sound reasonable?

**Comment by kaitejohnson** [comment]
> In short: I support removing the namespace qualifications from this PR (e.g. revert `ggplot2::ggplot()` to `ggplot()`), but keeping the universal use of the `.data` pronoun. @kaitejohnson, @gvegayon et al does that sound reasonable?

Yes I agree. I would also support importing a few other ggplot functions like `geom_line()`, `geom_point()`, `geom_vline/hline()` etc. but again this isn't super critical because I don't intend this package will be doing a ton of plotting anyway.

File: NAMESPACE
Comment (gvegayon) [reply NAMESPACE:74]: I think we can!

**Comment by gvegayon** [comment]
Thanks all for the comments. I will add `ggplot2`'s function to imports and keep the `.data` syntax.

File: R/wwinference.R
Comment (gvegayon) [inline R/wwinference.R:124]: @kaitejohnson, I think this fixes a bug. Before, it was getting `params` from somewhere I was not sure. Can you validate that what I did is correct?

File: R/generate_simulated_data.R
Comment (gvegayon) [reply R/generate_simulated_data.R:187]: I am not following what the need of the double bang operator is here.

File: R/generate_simulated_data.R
Comment (gvegayon) [reply R/generate_simulated_data.R:421]: Good point, fixing.

File: R/wwinference.R
Comment (kaitejohnson) [reply R/wwinference.R:124]: Oh yup that was a bug that probably hasn't gotten caught in the vignette because we defined `params` in the vignette.... 

Also realizing this is out of date we need to merge #58!

File: R/get_draws_df.R
Comment (gvegayon) [reply R/get_draws_df.R:80]: All mutates now use `nameofvariable = ` instead of `"nameofvariable"`.

File: R/get_stan_data.R
Comment (gvegayon) [reply R/get_stan_data.R:72]: Is this really needed?

File: R/get_stan_data.R
Comment (gvegayon) [reply R/get_stan_data.R:73]: Same here

File: R/get_stan_data.R
Comment (gvegayon) [reply R/get_stan_data.R:242]: This seems like a big change; am I missing something? cc @kaitejohnson

File: R/get_draws_df.R
Comment (gvegayon) [inline R/get_draws_df.R:59]: @dylanhmorris, this change breaks `pkgdown` but not `R CMD check`. Any idea why?

File: R/get_stan_data.R
Comment (kaitejohnson) [reply R/get_stan_data.R:242]: No so basically I just want the unique values of the total population column in the input count data here. 

I tend to do this with `select` for the column name and then `pull()` to get the vector, and then `unique()`. I often forget that this can be done more succinctly with `distinct(col_name)` and `pull()` bc distinct automatically does the selecting.

File: R/get_draws_df.R
Comment (dylanhmorris) [inline R/get_draws_df.R:137]: Just a missing exclamation point @gvegayon 
```suggestion
    tidybayes::spread_draws(!!str2lang("r_site_t[site_index, t]")) |>
```

File: R/get_draws_df.R
Comment (dylanhmorris) [reply R/get_draws_df.R:59]: Based on the logs, it looks like it's not the one you highlighted, it's this one: https://github.com/CDCgov/ww-inference-model/pull/108/files#r1737876278

so I think it's just the missing exclamation point.

**Comment by seabbs** [comment]
> .data[[] and string calls


It would be best if having f2f calls to put the resulting conversation/justifications explicitly in the PR notes or it can make decision-making hard to track. 

I assume there you mean their recommendation to use `.data$var` and not `.data[[var]]` which they don't make a strong case for that I can see.  From the discussion that is here it seems a bit that those things might be conflated (i.e if we are really following `dplyr` standards we should be supporting NSE as the tool for specifying input in our functions.

**Comment by kaitejohnson** [comment]
Here's how I would summarize the discussion: 

- tidyeval/`dplyr` error prone because of ambiguity with column names versus variables being passed to functions.
- R packages can look ugly/difficult to read under the hood. I personally am strongly in favor of readability, especially at this early stage, for those that would want to use this as I think it encourages contributions and makes it more approachable for someone to understand what a function is doing. 
- Using `.data[[var]]` whenever a column is being assigned, and using `.data$var` when something is being referenced, and using `.data[["var_name"]] whenever calling on a specific column name, and the `!!` whenever something is being assigned that is a variable not a column name....  (@gvegayon @dylanhmorris please correct me if I'm wrong/perhaps in this PR we could just clarify that so we know going forward what the standards are) 

My takeaway:
- package should have never been built using `dplyr` from the start given this. However, this would have made development a bit slower (for me personally) since despite being a modeling package, there is still a fair bit of data wrangling that needs to happen to format things properly with the two datasets

**Comment by seabbs** [comment]
>  error prone because of ambiguity with column names versus variables being passed to functions.

Is there evidence for this assertion? 

> R packages can look ugly/difficult to read under the hood

They can also not right.

> package should have never been built using dplyr f

Interestingly we will be switching `epidist` over from `data.table` to `dplyr` as it makes it much easier to get contributions. 

> Using

Sounds like a clear plan. Should add this steer to the contributing guide.

**Comment by gvegayon** [comment]
>> error prone because of ambiguity with column names versus variables being passed to functions.
>
> Is there evidence for this assertion?

The first thing I can think of is this issue with masking/unmasking caught by CFA in the Rt pipeline: [Teams discussion](https://teams.microsoft.com/l/message/19:63572bdbd6ec4f6080d3c103fa2ba2f5@thread.tacv2/1707942749024?tenantId=9ce70869-60db-44fd-abe8-d2767077fc8f&groupId=4f34eba7-8fcb-4a26-9b34-a434ea777f0c&parentMessageId=1707942749024&teamName=CFA-Predict&channelName=Topic%20-%20Dev%20-%20R&createdTime=1707942749024). I was able to reproduce the bug with had to do with NSE [here](https://github.com/cdcent/cfa-predict-sandbox/blob/2efa71ffb6e867bae7f061c12afd26223774316c/vegayon-xrd4/2024-02-14-scoping.md)

File: R/generate_simulated_data.R
Comment (dylanhmorris) [reply R/generate_simulated_data.R:189]: `ot` and `nt` are variables, so they should be evaluated. There _shouldn't_ be clashing column names but should is a dangerous word, so I always like to have variables on the RHS of `tidyeval` expressions evalated explicitly

**Comment by dylanhmorris** [comment]
> Using .data[[var]] whenever a column is being assigned, and using .data$var when something is being referenced, and using .data[["var_name"]] whenever calling on a specific column name, and the !! whenever something is being assigned that is a variable not a column name.... (@gvegayon @dylanhmorris please correct me if I'm wrong/perhaps in this PR we could just clarify that so we know going forward what the standards are)

This is close but not exact. As @seabbs says it may be simplest to follow the conventions in the `dplyr` vignette, particularly
https://dplyr.tidyverse.org/articles/in-packages.html#data-masking-and-tidy-selection-notes

There are a few places it is silent, however, and we can and should just choose a convention and be consistent

File: R/figures.R
Comment (gvegayon) [reply R/figures.R]: Addressed in 7b24e012982136dc989699cde5bf3db7d64a969a

File: R/get_stan_data.R
Comment (gvegayon) [inline R/get_stan_data.R:1]: Adding these changes broke R CMD check, @dylanhmorris

**Comment by kaitejohnson** [comment]
> Interestingly we will be switching epidist over from data.table to dplyr as it makes it much easier to get contributions.

I think that makes sense. Curious if you all use `autoglobals()` or follow the `dplyr` guidelines? 

This also points out that we need to write up contributing guidelines...

File: R/get_stan_data.R
Comment (dylanhmorris) [reply R/get_stan_data.R:1]: I think it's the `distinct |> pull()`, not the `!!`

```R
Error in `sum(ww_values$pop_ww)/pop`: non-numeric argument to binary operator
```

Comment (dylanhmorris) [reply R/get_stan_data.R:1]: Learned something new! The issue is that `distinct` isn't tidyselect under the hood.

```R
df <- tibble::tibble(a = c(1,2,3), b = c(1, 1, 2))

df |> dplyr::distinct("b")
```
```R
# A tibble: 1 × 1
  `"b"`
  <chr>
1 b
```

whereas

```R
df |> dplyr::distinct(data$b)
```
yields the expected result:
```R
# A tibble: 2 × 1
      b
  <dbl>
1     1
2     2
```

File: R/get_stan_data.R
Comment (dylanhmorris) [inline R/get_stan_data.R:242]:
```suggestion
    dplyr::distinct(.data$total_pop) |>
```

**Review by dylanhmorris — COMMENTED** [review]
Fixing the bug I introduced, apologies @gvegayon

File: R/get_draws_df.R
Comment (dylanhmorris) [inline R/get_draws_df.R:16]: This can now be deleted, no?

File: R/get_draws_df.R
Comment (dylanhmorris) [reply R/get_draws_df.R]: Remove these now that we've used the `sym2lang` trick?

**Comment by seabbs** [comment]
> Curious if you all use autoglobals() or follow the dplyr guidelines?

Autoglobals for the left hand side and nse for the right as needed I think

**Comment by kaitejohnson** [comment]
Btw since I presume this will be merged pretty close to as is and I expect it will lead to many git conflicts if I don't branch off if it, I am going to work from this PR. I would approve and merge but I see a number of unresolved conversations and I don't want to commit those suggestion myself. Looks to me like they are mostly choice of syntax.

File: R/generate_simulated_data.R
Comment (dylanhmorris) [inline R/generate_simulated_data.R:190]: Not for now, but for future reference this can be `dplyr::pull("date")`

File: R/get_stan_data.R
Comment (dylanhmorris) [inline R/get_stan_data.R:17]:
```suggestion
      .data$date > last_count_data_date - lubridate::days(!!calibration_time)
```

File: R/preprocessing.R
Comment (dylanhmorris) [inline R/preprocessing.R:187]:
```suggestion
      log_conc = log(.data[[conc_col_name]]),
```

**Review by dylanhmorris — CHANGES_REQUESTED** [review]
A couple very final things @kaitejohnson

File: R/generate_simulated_data.R
Comment (kaitejohnson) [reply R/generate_simulated_data.R:190]: Can you explain why it should be here -- is it because it is a column name and we should always refer to column names by "col_name" if they are not variables for column names?

Comment (dylanhmorris) [reply R/generate_simulated_data.R]: "Should" might be strong, but the reason it _can_ goes back to the distinction between "tidy selection" and "data masking". `dplyr::select()` and `dplyr::pull()`, among others, use[^1] tidy selection; functions like `dplyr::mutate()` and `dplyr::summarise()` use[^1] data-masking.

[^1]: For positional arguments after the first argument, which is typically the data frame to be manipulated

See https://dplyr.tidyverse.org/articles/programming.html#introduction

Comment (dylanhmorris) [reply R/generate_simulated_data.R]: The result is that 

```R
df |> dplyr::select("a_string")
```
means "select the column named `'a_string'` if it exists"

whereas:

```R
df |> dplyr::mutate(new_col = "a_string")
```
is interpreted as "create a new column named `'new_col'` whose value in each row is the string `'a_string'`. 

And if what you want is "create a new column named `'new_col'` whose values are the values of another column named `'a_string'`, you need:

```R
df |> dplyr::mutate(new_col = .data[["a_string"]])
```

or 

```R
df |> dplyr::mutate(new_col = .data$a_string)
```

**Review by dylanhmorris — APPROVED** [review]
LGTM, thanks @gvegayon / @kaitejohnson!

File: R/generate_simulated_data.R
Comment (kaitejohnson) [reply R/generate_simulated_data.R]: Or:
```r
df |> dplyr::mutate("new_col" = .data$a_string
```
Which I believe is the syntax we're rolling with here correct (we should always explicitly use the `""` for creating a new column name)?

**Pull request merged**
